crates/sui-core/src/execution_engine.rs

```diff
@@ -319,59 +358,34 @@ fn transfer_object<S>(
     Ok(())
 }
 
-/// Debit `coins` to pay amount[i] to recipient[i]. The coins are debited from left to right.
-/// A new coin object is created for each recipient.
-fn pay<S>(
-    temporary_store: &mut TemporaryStore<S>,
-    coin_objects: Vec<Object>,
-    recipients: Vec<SuiAddress>,
-    amounts: Vec<u64>,
-    tx_ctx: &mut TxContext,
-) -> Result<(), ExecutionError> {
+fn check_coins(
+    coin_objects: &[Object],
+    mut coin_type: Option<StructTag>,
```

Comment:
I am not following, it needs to be assigned if the input coin_type is None?

Reply:
Perhaps I can ask a different question: why do you need to return the type in the function return value if you are already assigning it through a mut ref argument?

Reply:
iiuc this line moves the `GasCoin::type_()` value instead of passing the mutable reference?
```
let (mut coins, coin_type) = check_coins(coin_objects, Some(GasCoin::type_()))?;
```
I cannot do `Option<&mut StructTag>` either as the None value will not be updated in `check_coins`, did I miss sth? can you elaborate the idiomatic way? thanks!

Reply:
Ah sorry I misread!

Reply:
all good! anything else I should address?